Fixes Issue 224 (Creating a bill generates a new unit every time)

kvotolato3 opened the pull request:
This pull request makes sure that bills are assigned to the existing unit, unless the user is a guest, in which case it will delete the old unit and bill and reassign new ones.
It also hides the "Number of Occupants" field when a logged in user already has a unit assigned to them.
Fixes issue #224

kvotolato3 committed:
Fixes issue #224 (Adding new bill generates new unit every time)

Added more case logic to bill creation. If user is guest, and has unit, existing bill and unit gets deleted, and new ones are created. This means guests only have one bill at a time.

Hid the number of occupants from form if user logged in and has a unit.

kvotolato3 committed:
Merge branch 'master' into issue_224

kvotolato3 committed:
Merge branch 'master' into issue_224

kvotolato3 committed:
Fix bill creation specs

kvotolato3 commented:
Please don't merge this yet. I noticed the "logging_in" method isn't running after a user logs in. Not sure if it has something to do with my changes or if it was there to begin with but, it should be taken care of.

kvotolato3 committed:
Changed "and" to "&&" in bills controller

kvotolato3 committed:
In teams#leaderboard (controller), used current_or_guest_user rather than current_user, since it automatically triggers the logging_in method when logging in for the first time. This step is essential for converting guests into full-fledged user accounts.

kvotolato3 commented:
This should be all set now!

mzagaja commented:
This mostly works, which is good. However when I created a new user account using Google it did not create a building for the unit. This then lead the following issue when I went to edit my user unit:

<img width="1060" alt="screen shot 2016-08-22 at 9 18 22 pm" src="https://cloud.githubusercontent.com/assets/565647/17877094/429110ae-68ae-11e6-9879-f4b05216ecc3.png">

To be honest at this point I would be ok with eliminating the concept of buildings at all and just using units. We can always add buildings back in the back end later.

kvotolato3 committed:
Merge branch 'master' into issue_224

kvotolato3 committed:
Fixed test 'user can create new bill from the team leaderboard'

- Removed occupants, because this field is hidden when number of occupants is already filled out on the user (our user factory automatically creates a unit with 1 occupant specified).

mzagaja commented on app/models/user.rb:
```diff
@@ -103,8 +103,35 @@ def decline_invite
-      email.split('@').first
+  def process_unit(unit_params, user_is_guest)
+    if user_is_guest # if user is guest
+      if unit = self.unit # and has a unit
```
what about `if user_is_guest && unit == self.unit`. Not sure if assignment is what you want to check here?

kvotolato3 replied:
Indeed I am checking an assignment here. Line 108 is saying "If we succeed in assigning self.unit to unit" (true), then do some things with that unit including destroy it, then create and assign a new unit. It is also implicitly saying "if we don't succeed in assigning self.unit to unit" (false), then just create and assign a new unit.

mzagaja replied:
Makes sense!

mzagaja replied:
However I think you should still be able to refactor to one line `user_is_gues && unit = self.unit`

kvotolato3 replied:
Yes, I understand.  Here is what it looks like using your suggestion. I had to add more lines and it ended up being the exact same number of lines altogether. What am I missing? I admit it looks repetitive, but I haven't been able to refactor this while keeping the desired behavior.

```
def process_unit(unit_params, user_is_guest)
    if user_is_guest && unit = self.unit # if user is guest and has a unit
      bill = Bill.find_by(unit: unit)
      bill.delete if bill != nil # delete guest's previously entered bills
      self.update(unit: nil) # delete guest's previously entered unit
      unit.destroy
      return self.assign_new_unit(unit_params)
    elsif user_is_guest #if user is guest and does not have a unit
      return self.assign_new_unit(unit_params)
    elsif unit = self.unit # user is not guest and has unit
      return unit
    else # user is not guest and no unit
      return self.assign_new_unit(unit_params)
    end
end
```

mzagaja commented on app/views/bills/_form.html.erb:
```diff
@@ -22,11 +34,13 @@
-      <%= units_form.label :number_occupants, "Number of Occupants in Your Home or Apartment" %><br>
-      <%= units_form.text_field :number_occupants, :id => "occupants" %>
-    </div>
+  <% if @show_occupants %>
```
Could you just do `unless user_signed_in? && current_user.unit.present?` here instead of having to do that in the controller and then passing it to a new variable?

kvotolato3 replied:
That's how I did it originally, but in this case I decided to move the tricky conditional into the controller so the view would be more readable and have less processing to do. I just added another commit that refactors the variable assignment in the controller to 1 line, which may help. If you still want me to, I can definitely move this into the view - I'm just going off the general practice of keeping as little logic as possible in the view.

mzagaja replied:
I think overall it isn't great to have to have that logic in the controller twice (both in `create` and `new`) as it is less DRY. Furthermore Sandi Metz suggests that more than one instance ("@") variable in the controller is not great (see also https://github.com/thoughtbot/guides/tree/master/best-practices#rails). To shorten it I wonder do we need to check for a signed_in user at all? Can it be sufficient to simply check for current_user.unit.present?. The documentation for .try might also yield a useful pattern here: http://apidock.com/rails/v4.2.1/Object/try.

Anyways some of this is merely trade-offs so I don't want to hold this up, but wanted to provide some helpful patterns to think about/try. Let me know if these suggestions make sense but if not happy to take it as is.

kvotolato3 replied:
No I totally agree with not having the instance variable being declared twice...that slipped by me last night and I realized it today. I will move the conditional into the view and see if we can use current_user.unit.present?

mzagaja reviewed (changes requested):
Looks well overall just had a couple questions and a suggestion.

kvotolato3 committed:
Refactored unless statement to a 1-liner in bills controller

mzagaja commented:
Not sure if it notifies you but I added additional comments to the line comments.

kvotolato3 committed:
Moved conditional logic into the create bill view

kvotolato3 commented:
Thanks for the suggestions! I just got rid of the "@show_occupants" variable and now we have _unless current_user.try(:unit).present?_ in the view. Seems to be working exactly as it did before, and is much cleaner.

kvotolato3 commented:
By the way, I've been leaving line comments as well. Still stuck on refactoring that "process_unit" method in user.rb. I tried your suggestions, but it ended up being the same number of lines in total.

kvotolato3 committed:
Refactored "process_unit" method in user.rb

kvotolato3 commented:
Just refactored a bit. @mzagaja Let me know what you think!

mzagaja reviewed (approved):
Looks better!
